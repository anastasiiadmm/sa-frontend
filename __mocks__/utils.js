import * as reduxHooks from 'react-redux';

jest.mock('../src/redux/store');
jest.mock('react-redux');

export const mockedUseSelectors = jest.spyOn(reduxHooks, 'useSelector');
export const mockedDispatch = jest.spyOn(reduxHooks, 'useDispatch');
