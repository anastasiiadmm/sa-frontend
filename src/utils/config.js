export const apiURL = '';

export const defaultError = { detail: 'На сервере что-то пошло не так' };
