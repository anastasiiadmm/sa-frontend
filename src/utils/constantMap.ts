import L from 'leaflet';

import active from 'assets/images/icons/active_tracktor.svg';
import endB from 'assets/images/icons/endB.svg';
import inactive from 'assets/images/icons/inactive_tracktor.svg';
import startA from 'assets/images/icons/startA.svg';

export const duckIconStart = new L.Icon({
  iconUrl: startA,
  iconRetinaUrl: startA,
  iconAnchor: new L.Point(14, 14),
  popupAnchor: new L.Point(14, 0),
  iconSize: new L.Point(30, 30),
  className: 'leaflet-div-icon',
});

export const duckIconEnd = new L.Icon({
  iconUrl: endB,
  iconRetinaUrl: endB,
  iconAnchor: new L.Point(14, 14),
  popupAnchor: new L.Point(14, 0),
  iconSize: new L.Point(30, 30),
  className: 'leaflet-div-icon',
});
export const activeIcon = new L.Icon({
  iconUrl: active,
  iconRetinaUrl: active,
  iconAnchor: new L.Point(14, 14),
  popupAnchor: new L.Point(14, 0),
  iconSize: new L.Point(28, 30),
  className: 'leaflet-div-icon',
});

export const inactiveIcon = new L.Icon({
  iconUrl: inactive,
  iconRetinaUrl: inactive,
  iconAnchor: new L.Point(14, 14),
  popupAnchor: new L.Point(14, 0),
  iconSize: new L.Point(28, 30),
  className: 'leaflet-div-icon',
});
