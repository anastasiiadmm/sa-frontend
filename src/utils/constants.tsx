export const pathsWithoutMargin = [
  '/open-map',
  '/field-climate',
  '/field-climate/station',
  '/field-climate/config',
];

export const climateOptions = [
  { value: 'last_communication', label: 'Последнее соединение' },
  { value: 'airTemp', label: 'Температура воздуха' },
  { value: 'soilTemp', label: 'Температура почвы' },
  { value: 'rh', label: 'Относительная влажность' },
  { value: 'rain24h', label: 'Осадки, 24ч' },
  { value: 'rain48h', label: 'Осадки, 2сут.' },
  { value: 'rain7d', label: 'Осадки, 7сут.' },
  { value: 'battery', label: 'Батарея' },
];

export const climateColors = {
  red: '%23ca3a3c',
  green: '%23107934',
  skyblue: '%231babe1',
  yellow: '%23f6c12d',
  orange: '%23ee741c',
  blue: '%230b75b9',
  gray: '%23d3d3d3',
};
