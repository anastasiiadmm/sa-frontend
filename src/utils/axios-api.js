import axios from "axios";
import {apiURL} from "utils/config";

export const axiosApi = axios.create({
  baseURL: apiURL,
});