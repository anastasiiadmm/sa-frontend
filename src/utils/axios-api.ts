import axios from 'axios';

import { checkForTokens, logoutUser } from 'redux/auth/authSlice';
import { apiURL } from 'utils/config';
import { logoutLocalStorage } from 'utils/token';
import store from 'redux/store';

const axiosApi = axios.create({
  baseURL: apiURL,
});

axiosApi.interceptors.request.use(async (config) => {
  const key = store.getState()?.auth?.tokens?.access;

  if (key) {
    config.headers = {
      ...config.headers,
      Authorization: `Bearer ${key}`,
      'Content-Type': 'application/json',
    } as any;
  }
  return config;
});

axiosApi.interceptors.response.use(
  async (config) => {
    return config;
  },
  async (error) => {
    const originalRequest = error.config;

    const statusCode = error?.response?.status;
    const index = store;
    const { tokens } = index.getState().auth;
    if (
      tokens &&
      statusCode === 401 &&
      error.config &&
      !error.config._isReady &&
      error.response.data.messages
    ) {
      originalRequest._isReady = true;
      try {
        const resp = await axiosApi.post('/accounts/refresh/', { refresh: tokens.refresh });
        if (resp.status === 200) {
          const newTokens = resp.data;
          axiosApi.defaults.headers.Authorization = `Bearer ${newTokens.access}`;
          const usersLocal = {
            user: index.getState()?.auth?.user,
            token: {
              access: resp.data.access,
              refresh: tokens.refresh,
            },
          };
          index.dispatch(checkForTokens(usersLocal));
          const obj: any = usersLocal;
          delete obj.token.refresh;
          localStorage.setItem('users', JSON.stringify(obj));
          window.dispatchEvent(new Event('storage'));
          return axiosApi(originalRequest);
        }
      } catch (e) {
        logoutLocalStorage();
        window.location.reload();
      }

      store.dispatch(logoutUser());
    }

    return Promise.reject(error);
  },
);

export default axiosApi;
