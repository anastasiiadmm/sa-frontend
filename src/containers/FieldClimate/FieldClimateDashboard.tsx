import bem from 'easy-bem';
import React, { useEffect, useState } from 'react';

import FieldClimateOpenMapComponent from 'components/FieldClimateOpenMapComponent/FieldClimateOpenMapComponent';
import FormField from 'components/FormField/FormField';
import Spinner from 'components/Spinner/Spinner';
import { useAppDispatch, useAppSelector } from 'redux/hooks';
import { fetchStations, stationsSelector } from 'redux/stations/stationsSlice';
import { MarkerData } from 'types/stationTypes';
import { climateOptions } from 'utils/constants';
import 'containers/FieldClimate/_fieldClimateDashboard.scss';

const FieldClimateDashboard = () => {
  const b = bem('FieldClimateDashboard');
  const dispatch = useAppDispatch();
  const { stations, stationsLoading } = useAppSelector(stationsSelector);
  const [markers, setMarkers] = useState<MarkerData[]>([]);
  const [selectedOption, setSelectedOption] = useState<string>(climateOptions[4].value);

  useEffect(() => {
    dispatch(fetchStations());
  }, []);

  useEffect(() => {
    if (stations && Array.isArray(stations)) {
      const transformedData = stations?.map((item) => ({
        id: item?.name?.original,
        position: item?.position?.geo?.coordinates,
        name: item?.name?.custom,
        dates: item?.dates,
        meta: item?.meta,
      }));
      setMarkers(transformedData);
    }
  }, [stations]);

  const handleChange = (value: string) => {
    setSelectedOption(value);
  };

  return (
    <React.Suspense fallback={<Spinner />}>
      {stationsLoading && <Spinner />}
      {!stationsLoading && (
        <FieldClimateOpenMapComponent markers={markers} selectedOption={selectedOption} />
      )}
      <div className={b('')}>
        <div>
          <FormField
            type='select'
            defaultValue={climateOptions[4]}
            handleChange={handleChange}
            options={climateOptions}
          />
        </div>
      </div>
    </React.Suspense>
  );
};

export default FieldClimateDashboard;
