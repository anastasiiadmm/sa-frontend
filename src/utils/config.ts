import { AxiosRequestConfig, AxiosRequestHeaders } from 'axios';

import { checkForTokens, logoutUser } from 'redux/auth/authSlice';
import store from 'redux/store';
import { TSetupAxiosInterceptor } from 'type';

import { apiUrlsV2, serverUrls, serverUrlsSocket } from 'utils/constants';
import { logoutLocalStorage } from 'utils/token';

const { REACT_APP_ENVIRONMENT } = process.env;

const appEnvironment = REACT_APP_ENVIRONMENT || 'local';

export const apiURL2 = apiUrlsV2[appEnvironment];
export const socketApiSocket = serverUrlsSocket[appEnvironment];

export const apiUrlCrop = serverUrls[appEnvironment];

export const setupAxiosInterceptors: TSetupAxiosInterceptor = (axiosInstance) => {
  return axiosInstance.interceptors.request.use(async (config: AxiosRequestConfig) => {
    const key = store.getState()?.auth?.tokens?.access;

    if (key) {
      config.headers = {
        ...config.headers,
        Authorization: `Bearer ${key}`,
      } as AxiosRequestHeaders;
    }
    return config;
  });
};

export const setupAxiosInterceptorsResponse: TSetupAxiosInterceptor = (axiosInstance) => {
  return axiosInstance.interceptors.response.use(
    async (config) => {
      return config;
    },
    async (error) => {
      const originalRequest = error.config;

      const statusCode = error?.response?.status;
      const index = store;
      const { tokens } = index.getState().auth;
      if (
        tokens?.access &&
        statusCode === 401 &&
        error.config &&
        !error.config._isReady &&
        error.response.data.messages
      ) {
        originalRequest._isReady = true;
        try {
          const resp = await axiosInstance.post('/accounts/refresh/', { refresh: tokens.refresh });
          if (resp.status === 200) {
            const newTokens = resp.data;
            axiosInstance.defaults.headers.Authorization = `Bearer ${newTokens.access}`;
            const usersLocal = {
              access: resp.data.access,
              refresh: tokens.refresh,
              is_manager: tokens.is_manager,
            };
            index.dispatch(checkForTokens(usersLocal));
            const obj = {
              access: resp.data.access,
              is_manager: tokens.is_manager,
            };
            localStorage.setItem('users', JSON.stringify(obj));
            window.dispatchEvent(new Event('storage'));
            return axiosInstance(originalRequest);
          }
        } catch (e) {
          logoutLocalStorage();
          window.location.reload();
        }

        store.dispatch(logoutUser());
      }

      return Promise.reject(error);
    },
  );
};
