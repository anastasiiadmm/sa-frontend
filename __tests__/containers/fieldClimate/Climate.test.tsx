import { BrowserRouter } from "react-router-dom";
import { cleanup } from "@testing-library/react";
import "../../../__mocks__/matchMedia.mock";
import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { act } from "react-test-renderer";
import { mockedDispatch, mockedUseSelectors } from "../../../__mocks__/utils";
import FieldClimateDashboard from "../../../src/containers/FieldClimate/FieldClimateDashboard";

afterEach(cleanup);

jest.mock('react-leaflet', () => ({
  MapContainer: jest.fn(),
  Marker: jest.fn(),
  Tooltip: jest.fn(),
  Popup: jest.fn(),
  TileLayer: jest.fn()
}));

describe("<FieldClimateDashboard />", () => {
  test("FieldClimateDashboard map component should be in the document", async () => {
    mockedUseSelectors.mockReturnValue({});
    const dispatch = jest.fn();
    mockedDispatch.mockReturnValue(dispatch);

    let tree;
    await act(async () => {
      tree = renderer
        .create(
          <BrowserRouter>
            <FieldClimateDashboard />
          </BrowserRouter>
        )
        .toJSON();
    });

    expect(tree).toMatchSnapshot();
  });
});
