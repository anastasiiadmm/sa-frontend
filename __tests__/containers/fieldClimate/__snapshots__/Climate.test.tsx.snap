// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldClimateDashboard /> FieldClimateDashboard map component should be in the document 1`] = `null`;
